modules/tribes/tests/client/components/TribesPage.component.tests.js

```diff
@@ -0,0 +1,202 @@
+import React from 'react';
+import { within } from '@testing-library/react';
+import { render, fireEvent, waitForElement } from '@testing-library/react';
+import '@testing-library/jest-dom/extend-expect';
+import '@/config/client/i18n';
+
+import TribesPage from '@/modules/tribes/client/components/TribesPage.component';
+import * as tribesApi from '@/modules/tribes/client/api/tribes.api';
+
+const api = { tribes: tribesApi };
+
+jest.mock('@/modules/tribes/client/api/tribes.api');
+
+const onMembershipUpdated = jest.fn();
+
+const dummyTribes = [{
+  _id: 'aaaa',
+  id: 'aaaa',
+  count: 20000,
+  slug: 'hitchhikers',
+  label: 'Hitchhikers',
+}, {
+  _id: 'bbbb',
+  id: 'bbbb',
+  count: 5000,
+  slug: 'snails',
+  label: 'Snails',
+}, {
+  _id: 'cccc',
+  id: 'cccc',
+  count: 1500,
+  slug: 'teddybears',
+  label: 'Teddy Bears',
+}];
+
+const dummyUser = {
+  _id: 'user0',
+  id: 'user0',
+  memberIds: ['cccc', 'aaaa'],
+};
+
+api.tribes.read.mockImplementation(async () => dummyTribes);
+api.tribes.join.mockImplementation(async id => ({ tribe: dummyTribes.find(tribe => tribe._id === id) }));
+api.tribes.leave.mockImplementation(async id => ({ tribe: dummyTribes.find(tribe => tribe._id === id) }));
+
+// helper function to determine a membership
+const isMember = (user, tribe) => user.memberIds.includes(tribe._id);
+
+// TextMatch function for finding an element with arbitrarily nested text
+// https://testing-library.com/docs/dom-testing-library/api-queries#textmatch
+const nestedTextMatch = text => (content, element) => element.textContent.includes(text);
+
+// helper that renders the TribesPage, waits for the tribes to be loaded from API and returns the rendered page
+const renderAndWaitForTribes = async ({ user }) => {
```

Comment:
What do you think about extracting the render + tribe-waiting to `beforeEach` like this? Does it make the code drier, or harder to read?

Reply:
Definitely DRYer. 👍